packages/nuxt/src/components/runtime/client-delayed-component.ts

```diff
@@ -0,0 +1,175 @@
+      return () => h(comp, mergeProps(attrs, { 'data-allow-mismatch': '', 'onVnodeMounted': hydrated }))
+    },
+  })
+}
+
+/* @__NO_SIDE_EFFECTS__ */
+export const createLazyNetworkComponent = (loader: AsyncComponentLoader) => {
+  return defineComponent({
+    inheritAttrs: false,
+    props: {
+      hydrate: {
+        type: Number,
+        required: false,
+      },
+    },
+    emits: ['hydrated'],
+    setup (props, { attrs, emit }) {
+      const hydrated = () => { emit('hydrated') }
+      if (props.hydrate === 0) {
+        const comp = defineAsyncComponent(loader)
+        return () => h(comp, mergeProps(attrs, { 'data-allow-mismatch': '', 'onVnodeMounted': hydrated }))
+      }
+      const comp = defineAsyncComponent({ loader, hydrate: hydrateOnIdle(props.hydrate) })
+      // TODO: fix hydration mismatches on Vue's side. The data-allow-mismatch is ideally a temporary solution due to Vue's SSR limitation with hydrated content.
+      return () => h(comp, mergeProps(attrs, { 'data-allow-mismatch': '', 'onVnodeMounted': hydrated }))
+    },
+  })
+}
+
+/* @__NO_SIDE_EFFECTS__ */
+export const createLazyEventComponent = (loader: AsyncComponentLoader) => {
+  return defineComponent({
+    inheritAttrs: false,
+    props: {
+      hydrate: {
+        type: [String, Array],
+        required: false,
+        default: 'mouseover',
+      },
+    },
+    emits: ['hydrated'],
+    setup (props, { attrs, emit }) {
+      const hydrated = () => { emit('hydrated') }
+      // @ts-expect-error Cannot type HTMLElementEventMap in props
+      const comp = defineAsyncComponent({ loader, hydrate: hydrateOnInteraction(props.hydrate) })
+      // TODO: fix hydration mismatches on Vue's side. The data-allow-mismatch is ideally a temporary solution due to Vue's SSR limitation with hydrated content.
+      return () => h(comp, mergeProps(attrs, { 'data-allow-mismatch': '', 'onVnodeMounted': hydrated }))
+    },
+  })
+}
+
+/* @__NO_SIDE_EFFECTS__ */
+export const createLazyMediaComponent = (loader: AsyncComponentLoader) => {
+  return defineComponent({
+    inheritAttrs: false,
+    props: {
+      hydrate: {
+        type: String,
+        required: false,
+        default: '(min-width: 1px)',
+      },
+    },
+    emits: ['hydrated'],
+    setup (props, { attrs, emit }) {
+      const hydrated = () => { emit('hydrated') }
+      const comp = defineAsyncComponent({ loader, hydrate: hydrateOnMediaQuery(props.hydrate) })
+      // TODO: fix hydration mismatches on Vue's side. The data-allow-mismatch is ideally a temporary solution due to Vue's SSR limitation with hydrated content.
+      return () => h(comp, mergeProps(attrs, { 'data-allow-mismatch': '', 'onVnodeMounted': hydrated }))
+    },
+  })
+}
+
+/* @__NO_SIDE_EFFECTS__ */
+export const createLazyIfComponent = (loader: AsyncComponentLoader) => {
+  return defineComponent({
+    inheritAttrs: false,
+    props: {
+      hydrate: {
```

Comment:
Is the prop name `hydrate` used on all of these components for consistency? Otherwise, I suggest changing the prop name to other name that are more "clear", for example, `condition` for `<LazyIfXxx>`, `promise` for `<LazyPromiseXxx>`, ...

Reply:
Yes, that is indeed what I was aiming for. It makes for a unified and easily-adoptable feature that people won't need to look around in the docs for the exact prop for each hydration strategy. The fact this allows to reuse a type to reduce the size of component IntelliSense is an added bonus.

Reply:
> Though I am curious, is there any issue that you foresee with this naming convention?

If any, I think that this name is too generic and maybe used by a feature of Vue in the future (though this should be rare, as Evan is aware of Nuxt).

Reply:
>  maybe used by a feature of Vue in the future

This did happen, but back then the prop was named as `loader` (which conflicted with the loader in `defineAsyncComponent`), which is actually why I renamed it to `hydrate`. I have to say, that the chances of Vue introducing something like that are very slim, especially because the delayed hydration system is pretty much stable now in terms of API (maybe not so much in terms of overall features). 

What name would you think would be better? Ideally it should be a shared name because of the consistency, which would bring a lot more benefit to newcomers.

Reply:
> What do you think about the name precondition?

This is definitely a unique name. May I ask why you think it's better? Hydrate at least has some context awareness as for its purpose, whereas precondition is pretty vague in this sense.

I'm not entirely opposed to it, I'm just curious to hear your opinion =)

Reply:
I've been thinking about this a little.

I do still believe that some variation of `hydrate(*?)` will be the best fit, because after all, it's for delayed hydration.
I also believe that keeping it uniform will make it much easier for newcomers to get along with it and learn about it, and it can also make it much more convenient to document.

I suppose `hydrateOn` is an equally decent name, but it sounds almost promise-condition related. What do you think?

Reply:
May I ask what is the logic behind using fake component prefixes instead of fake wrapper component that can take props to determine hydration method?

I fear that those prefixes may not be very portable, they might conflict with lint rules and prevent ide autocompletion from working properly

For example do you see any problem with this: `<TimerHydrate :time="3000"><MyComponent></TimerHydrate>` ?

We have control on the vue template loader in nuxt, so we could definitely transform this and avoid prop name conflicts in the first place, and this solves a lot of headaches

Reply:
Hey @Tofandel ! Thanks for voicing your concern. From your example, there is more boilerplate involved as you need to add more to your template. Compare that with the following:
`<LazyTimeMyComponent :hydrate="3000" />`

I'm not entirely sure what you mean by prevent IDE autocompleting from working properly and which lint rules they could conflict with, so if you could elaborate on that it would be appreciated

Reply:
Ide provide autocompletion on props name via the imported components in the Vue file or types of global components

So for the autocompletion to work that means all possible components names need to be globally typed, this is what you are already doing a bit lower right? 

Linters might try to understand if a component used exists and report errors if it doesn't

Ide provide autocompletion on props name via the imported components in the Vue file or types of global components

I'm just wondering if then having so many types added might be problematic (imagine I have 300 components, before they were there only twice right? The Lazy version and the normal version, 600 types is still okay, but what happens when now all 300 are multiplied 9 times with this PR?, seems like it would snowball quickly)

Maybe using a modifier on lazy components then could be a better solution like

<LazyMyComponent v-hydrate:timer="3000">

Because then we only need to provide types for the modifiers once and we still don't have the problem with the 'hydrate' and prop naming and we can correctly type each prop

Reply:
You bring some important points, but unfortunately they come with their own downsides.

While your approach could help with the amount of types, I can't help but to think about the performance impact.

Let's start with the obvious: it's very rare to have 300 components in an app, no matter the size. 20? Very much possible. 40? Sure. 70?? Maybe if it covers so many things and is extremely multi purpose. But 300 components seems like quite a jump. Furthermore, I don't believe having these 2700 types would be too much of a concern for people, regardless of where it's deployed, for 3 main reasons:
1. It's not bundled with the app so it doesn't add any additional bundle size.
2. It's written very infrequently and is only used in development, which usually happens on a developer's local computer or a dedicated server, which don't have a memory bottleneck.
3. If you take a look at all the built-in type declarations for every type in web development, whether it be NodeJS types, DOM types, and all others, you can see it adds up to way more than 2700 types, and yet it works and performs perfectly fine in development, which leads me to believe the impact from the types might not be that significant. 

Now let's go over the performance aspect:
While having a directive could make it seem better, this could actually slow it down.

In the current approach, each delayed hydration wrapper is made exclusively for a single purpose. This allows for maximal optimizations, as it only needs to take care of a single scenario.

In your approach, the handling of the `<Lazy*>` components would need to account for regular, non-delayed async components, as well as all possible delayed hydration strategies, all from the same place. This inherently makes it impossible to optimize for ideal performance as it needs to have broader handling that might result in something for a certain input and might result in something else for a different one.

Reply:
> In your approach, the handling of the <Lazy*> components would need to account for regular, non-delayed async components, as well as all possible delayed hydration strategies, all from the same place. This inherently makes it impossible to optimize for ideal performance as it needs to have broader handling that might result in something for a certain input and might result in something else for a different one.

You misunderstood slightly the approach, it would be only syntactic sugar directive (it would never work as a real directive), it would actually be transformed at build time by `@vue/sfc-compiler` or `vite` and during the transformation it would not be a problem to use a different wrapper component, this means all that you did with the wrappers stays mostly the same, it only needs a transformer and the type generation is not needed anymore, we could still let the wrappers be used via imports as well in case someones want to use a wrapper component outside of a template like a fully scripted component

It would not have any performance drawback compared to this method

Reply:
Ah, I see what you're referring to. At first you stated it would be transformed at runtime, which is where the confusion arose. Doing it at build time does indeed sound better but the directive also should be unique to ensure it doesn't conflict with other libraries (I think I recall some external library using `v-hydrate`).

> My 16Gb of rams would like to disagree, I'm constantly swapping when developing with a single project taking in the 10Gb of ram

Well, that's definitely odd. Obviously I can't outright say this is wrong, but I am simply speaking from experience with an experiment that added thousands of types, and all seemed to be fine and fast. My IDE didn't appear to be slower either

Reply:
My bad, definitely meant build time

> Doing it at build time does indeed sound better but the directive also should be unique to ensure it doesn't conflict with other libraries (I think I recall some external library using v-hydrate).

I think it shouldn't be a problem because it's a directive that we will only transform on a Lazy prefixed component, so even if naming conflicts with real directives, because of the context in which it's used is scoped to a specific nuxt feature and we will be transforming only those instances it shouldn't cause issues, and the other library might become redundant if they are providing the same feature as well

For performance it really depends on your env and your ide, they are not all created equal and some try to provide more features for the DX like AI autocompletion at the cost of performance or memory, it's a sensitive balance

In any case I think v-hydrate:how is also easier to remember and understand for the end users

Reply:
> Let's start with the obvious: it's very rare to have 300 components in an app, no matter the size. 100? Very much possible. 200? Sure. 250?? Maybe if it covers so many things and is extremely multi purpose. But 300 components doesn't seem like an everyday occurrence. Furthermore, I don't believe having these 2700 types would be too much of a concern for people, regardless of where it's deployed, for 3 main reasons:
> 

For what it's worth, I don't consider my app _**that**_ large, and yet...
![image](https://github.com/user-attachments/assets/e18b57c2-d0a3-4da7-be54-f8b6514c7f43)

Reply:
> In any case I think v-hydrate:how is also easier to remember and understand for the end users

This would probably vary. Ideally, as I've stated in the past, a uniform naming would make it easier for newcomers because they won't need to figure out which option is for which strategy.

> For what it's worth, I don't consider my app that large, and yet...

Well, that's definitely a rare occurrence, but I agree there are cases that do reach it, mainly with UI libraries.

All in all, I'd have to look into the plugin approach, especially from the side of the DX with the directive intellisense, because this would be a tricky aspect that I do not want to go unaddressed.

Reply:
> Well, that's definitely a rare occurrence, but I agree there are cases that do reach it, mainly with UI libraries.

I'm not using any UI libraries. That's just normal userland, custom Vue SFC components.

Reply:
> a uniform naming would make it easier for newcomers because they won't need to figure out which option is for which strategy.

Which is exactly what I was advocating v-hydrate:[name-of-strategy]="propValue" achieves, because then you don't have any option name involved at all

I do think a normal sized app (website with maybe a blog or ecommerce) easilly reaches the 200 or 300 count

This is me and it's also not a big app and it's still in development so likely there will be 50 more components by the end of it
![image](https://github.com/user-attachments/assets/335f3103-0d6e-4c87-b425-3d26f629aa28)

Reply:
> v-hydrate:[name-of-strategy]="propValue"

To clarify, this part still needs the unique identifier being the strategy. I meant uniform in the prop, which is why I initially opted for the prefixes with :hydrate, as people won't need to figure out which prop is the one for time-based hydration and which one is for promise-based hydration. In either case however, the actual strategy is still unique.

On the bright side, I've been able to make a prototype implementation using a build plugin as you suggested. A demonstration is provided below:
 
![image](https://github.com/user-attachments/assets/78a0c0dc-3170-4e61-ab3d-69b9bf6df268)

https://github.com/user-attachments/assets/ced2be81-8e53-4b8f-9c65-ffc64c1da7a5

Reply:
Amazing ❤️, let me know if you need help creating the type for the directive

Reply:
> I see you used a prop instead of a directive

For now, yes, because I've been looking into typing the directive. I'm also taking into account the warning not to use custom directives on components because it can have unexpected results with multiple root nodes, so there is something which I would need to carefully review.

Reply:
Hmm, I have noticed an issue with this. In the previous approach with the component prefixes, you could simply use the prefix without any other props for the default. However, if we rework this into a prop, we end up with:
```vue
<LazyMyComponent hydrate:time />
```

If we want to allow for defaults, this means the type would need to allow boolean for each hydration strategy, otherwise a type error will occur. While it's not a big deal, I wonder if it's the best approach for this

Reply:
@Tofandel I'm not entirely sure if this is what you were looking for, but feel free to review it and let me know if there's anything else you think I should change.

I can't fully say I am extremely pleased with the implementation because there are still edge cases that can't be fixed during the build process without overly impacting the performance (for example, message above), and I would still like to sit on these issues a little and try to see if I can fix them. It's much less straightforward than when it was prefix-based, that's for sure, and I am hesitant to use an AST traversal in the plugin because it's more expensive than regex, which for the most part should suffice for the general guards, but still more performant

Reply:
I'll have a look, but I'm quite surprised to see a regex in there, normally you already have full AST available if you build the plugin for the sfc compiler, I'll maybe see too send a PR to your fork to demonstrate

Reply:
Where do you see the full AST? When making such plugins you get the raw code as a string and would have to walk the AAT using acorn or another AST library

Reply:
Basically you need to pass into the Vue vite plugin a function into an option under compilerOptions.nodeTransforms

Reply:
Ah, that's where the issue lies I assume. Build plugins happen before the Vue plugin is ran. You can see that other plugins in the Nuxt codebase traverse the AST on their own with Acorn rather than relying on Vue

Reply:
You can create the vite plugin that simply edits the config of the other Vue plugin I think. Acorn is meant to parse javascript or ts, not vue templates

Reply:
Maybe, but this seems way too overcomplicated, and I'llhave to carefully test it.. I'm almost considering maybe reverting to the prefix approach and just refining it. We can make use of mapped types to easily type out all possible strats within a single type definitions which will massively reduce total types (it's actually used in here as well so it could be applied the same way). Something that would look roughly like this:

MyComponent: regular import,
LazyMyComponent: regular async import,
Lazy${k in keyof types}MyComponent: import with hydrate: ${typeMap[k]}

Would this not essentially answer your concern about total types?

Reply:
`Lazy${k in keyof types}MyComponent: import with hydrate: ${typeMap[k]}` is this possible to do with a top level type?

Just sit tight, I'll send in a PR to your branch tomorrow night with some tweaks (like userland extensibility of the hydration methods) and the AST improvement

Reply:
Hey!

After careful consideration, I decided to revert the directive approach and return to the original, for a couple main reasons:
1. The directive approach brings a lot of complexity to the process, and would make it more difficult to maintain and extend it. There could be some unaccounted edge cases and having another AST manipulation, in my opinion, isn't worth the performance degradation, as some projects can have tens of delayed hydration components at a time in a single page, which would be quite the overhead for the AST traversal.
2. There is no concrete evidence that suggests the types are a major bottleneck. Most IDE's are already optimized for type definitions, and Microsoft themselves are already optimizing TypeScript's internal overhead, and there is no indication that it would actually cause a problem that warrants a complete build time transformation.

That being said, feel free to still send in the PR, if you believe it is worth it, and of course this could be reworked again, once everything has been planned out, to avoid bringing more harm than good. I will still help you work on the PR, if you'd like.

However, there should be some additional testing and real-world cases to accurately determine whether the types cause too major of an issue.

Hopefully you're not too upset about my decision, and of course I would love to hear any insight you may have.

Reply:
> we can apply transforms to the default slot to make all the components inside lazy and import the slot on demand

I haven't really tried that, but do you think as it currently is won't allow for slots?

Reply:
Just letting you know, I cloned the original branch to patch-21-backup so we could continue working on it. In the meantime, I believe some functional implementation is better than no implementation, especially with the performance considerations the current approach takes care of. We could always make another PR to refactor the delayed hydration process, but I didn't want this to be a road block from implementing it.